feat(forms): add an input property to form directives that makes ngSub…

DibyodyutiMondal opened the pull request:
…mit wait for async validation to complete before it emits

fix #31021

Before this change, ngSubmit would emit immediately after running sync validation, even though async validation was still running. Thus, any code that handled ngSubmit had no guarantees that the form data was in fact valid.

After this change, adding `submitAfterAsyncValidation` to the form tag in the template will indicate the form directive to wait for async validation to complete before it emits.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

dylhunn reviewed (changes requested):
I'll have an in-depth look shortly. Would you please add tests as well, in the `.spec` files? Thanks!

Edit: One more thing -- let's make the commit message `feat` instead of `fix`.

Edit 2: looks like you also need a rebase:
```
git fetch upstream master
git checkout fix-async-validation-and-ngSubmit
git rebase upstream/master
git push --force-with-lease
```

AndrewKushnir reviewed (commented):
Just a quick comment: I think we should consider adding this as a flag to the `FormsModule` and `ReactiveFormsModule` instead ([see here](https://github.com/angular/angular/blob/master/packages/forms/src/form_providers.ts#L58)), so that users can configure it once for all forms (vs adding an extra property in templates for each [formGroup] and `<form>` tag).

AndrewKushnir commented:
@dylhunn I think we could also check (by running a TGP) if it can be enabled by default with an opt-out flag (for apps that might be broken because of that). WDYT?

DibyodyutiMondal commented:
> Just a quick comment: I think we should consider adding this as a flag to the `FormsModule` and `ReactiveFormsModule` instead

Makes a lot of sense. I'll try to put that in...

> @dylhunn I think we could also check (by running a TGP) if it can be enabled by default with an opt-out flag 

I'll keep the default as false at the moment, and leave it to you guys to change as you see fit later

> Would you please add tests as well, in the .spec files? Thanks!

Honestly, I needed help with those, to the point I gave up altogether. Please correct me if I am mistaken - we need the following tests:
- Template driven forms
  - should not wait for async validation by default
  - should wait for async validation if flag is true
- Reactive forms
  - should not wait for async validation by default
  - should wait for async validation if flag is true

DibyodyutiMondal commented:
Also, how do I build and test this on my local system, I managed to install the packages using yarn.
Running 'yarn test' gave this output:
```
INFO: Analyzed 0 targets (0 packages loaded, 0 targets configured).
INFO: Found 0 test targets...
INFO: Elapsed time: 121.259s, Critical Path: 0.01s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
error Command failed with exit code 4.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

dylhunn commented:
> Also, how do I build and test this on my local system, I managed to install the packages using yarn. Running 'yarn test' gave this output:
> 
> ```
> INFO: Analyzed 0 targets (0 packages loaded, 0 targets configured).
> INFO: Found 0 test targets...
> INFO: Elapsed time: 121.259s, Critical Path: 0.01s
> INFO: 1 process: 1 internal.
> INFO: Build completed successfully, 1 total action
> INFO: Build completed successfully, 1 total action
> error Command failed with exit code 4.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> ```

Try the following:
```
yarn bazel test packages/forms/test
```
If that doesn't work, let me know and I'll try to help!

dylhunn commented:
@DibyodyutiMondal

> Honestly, I needed help with those, to the point I gave up altogether. Also where should I put them? Under the test group for async validation, or under the group for 'submit'? Also, please correct me if I am mistaken - we need the following tests:
> 
> * Template driven forms
>   
>   * should not wait for async validation by default
>   * should wait for async validation if flag is true
> * Reactive forms
>   
>   * should not wait for async validation by default
>   * should wait for async validation if flag is true

Yes, that sounds like a really good list of tests! 

Try adding your new test to `test/template_integration_spec.ts`, inside the section `'submit and reset events'`. You can copy/paste `should mark NgForm as submitted on submit event` as a starting point. Once you have a test, you can run it with `yarn bazel test packages/forms/test`.

Regarding the test itself, as a hint, you can use `fakeAsync` and `tick` to simulate asynchronous code. Give it a try and let me know if you need more help. :)

DibyodyutiMondal commented:
```
yarn run v1.22.17
$ D:\Github\angular\node_modules\.bin\bazel test packages/forms/test
INFO: Analyzed target //packages/forms/test:test (1 packages loaded, 103 targets configured).
INFO: Found 1 test target...
ERROR: D:/github/angular/packages/localize/src/localize/BUILD.bazel:5:11: Compiling TypeScript (devmode) //packages/localize/src/localize:localize failed: Worker process did not return a WorkResponse:

---8<---8<--- Start of log, file at C:/users/dibyo/_bazel_dibyo/cpnblasi/bazel-workers/worker-14-TypeScriptCompile.log ---8<---8<---
/bin/bash: C:/users/dibyo/_bazel_dibyo/cpnblasi/execroot/angular/bazel-out/host/bin/external/npm/@bazel/typescript/bin/tsc_wrapped.sh: No such file or directory
---8<---8<--- End of log ---8<---8<---
Target //packages/forms/test:test failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: D:/github/angular/packages/forms/test/BUILD.bazel:28:18 Middleman _middlemen/packages_Sforms_Stest_Stest.bat-runfiles failed: Worker process did not return a WorkResponse:

---8<---8<--- Start of log, file at C:/users/dibyo/_bazel_dibyo/cpnblasi/bazel-workers/worker-14-TypeScriptCompile.log ---8<---8<---
/bin/bash: C:/users/dibyo/_bazel_dibyo/cpnblasi/execroot/angular/bazel-out/host/bin/external/npm/@bazel/typescript/bin/tsc_wrapped.sh: No such file or directory
---8<---8<--- End of log ---8<---8<---
INFO: Elapsed time: 1.312s, Critical Path: 0.61s
INFO: 10 processes: 10 internal.
FAILED: Build did NOT complete successfully
//packages/forms/test:test                                      FAILED TO BUILD

FAILED: Build did NOT complete successfully
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

DibyodyutiMondal commented:
If it's any help, I have a windows PC 64-bit

dylhunn commented:
Sorry to see you're having build issues. Can you try again with `--verbose_failures` and post the detailed output?

DibyodyutiMondal commented:
Still getting the same error. It seems to be expecting some file to exist. I even went and checked the log file it mentioned, but that just contains a copy of the message given here

```
yarn run v1.22.17
$ D:\Github\angular\node_modules\.bin\bazel test packages/forms/test --verbose_failures
INFO: Analyzed target //packages/forms/test:test (1 packages loaded, 87 targets configured).
INFO: Found 1 test target...
ERROR: D:/github/angular/packages/zone.js/lib/BUILD.bazel:9:11: Compiling TypeScript (devmode) //packages/zone.js/lib:zone_d_ts [for host] failed: Worker process did not return a WorkResponse:

---8<---8<--- Start of log, file at C:/users/dibyo/_bazel_dibyo/cpnblasi/bazel-workers/worker-30-TypeScriptCompile.log ---8<---8<---
/bin/bash: C:/users/dibyo/_bazel_dibyo/cpnblasi/execroot/angular/bazel-out/host/bin/external/npm/@bazel/typescript/bin/tsc_wrapped.sh: No such file or directory
---8<---8<--- End of log ---8<---8<---
Target //packages/forms/test:test failed to build
ERROR: D:/github/angular/packages/forms/test/BUILD.bazel:28:18 Middleman _middlemen/packages_Sforms_Stest_Stest.bat-runfiles failed: Worker process did not return a WorkResponse:

---8<---8<--- Start of log, file at C:/users/dibyo/_bazel_dibyo/cpnblasi/bazel-workers/worker-30-TypeScriptCompile.log ---8<---8<---
/bin/bash: C:/users/dibyo/_bazel_dibyo/cpnblasi/execroot/angular/bazel-out/host/bin/external/npm/@bazel/typescript/bin/tsc_wrapped.sh: No such file or directory
---8<---8<--- End of log ---8<---8<---
INFO: Elapsed time: 4.121s, Critical Path: 3.50s
INFO: 10 processes: 10 internal.
FAILED: Build did NOT complete successfully
//packages/forms/test:test                                      FAILED TO BUILD

FAILED: Build did NOT complete successfully
error Command failed with exit code 1.
```

dylhunn commented:
Interesting. Perhaps you could try the advice from here? [Github link](https://github.com/bazelbuild/rules_nodejs/issues/1880#issuecomment-627729432) That issue advices that you change the order in which cygwin loads on your path.

We *theoretically* support building Angular on Windows, but to the best of my knowledge, none of us on the framework team actually use a Windows PC. If definitely should work on Linux or Mac, if you have access to those platforms. If not, @josephperrott, have you seen anything like this before?

DibyodyutiMondal commented:
Ok, let's see if WSL helps

dylhunn commented:
This doesn't seem to be in a reviewable state -- let me know when you've been able to run the tests and get everything green. Thanks!

DibyodyutiMondal commented:
Each package tested generated a console log similar to this one
```
INFO: From NpmPackageBin packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_styling/component_styles/_generated.js:
(node:34) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./" in the "exports" field module resolution of the package at /home/dibyodyuti/.cache/bazel/_bazel_dibyodyuti/d4a32c8202ac3fa7a590bc2bd0fa04c7/sandbox/linux-sandbox/467/execroot/angular/bazel-out/k8-opt-exec-2B5CBBC6/bin/packages/compiler-cli/test/compliance/test_cases/partial_r3_view_compiler_styling/component_styles.sh.runfiles/angular/node_modules/tslib/package.json.
Update this package.json to use a subpath pattern like "./*".
(Use `node --trace-deprecation ...` to show where the warning was created)
```

A few changes that I would have appreciated being mentioned in the developer docs.

1. To run yarn install with the `--network-timeout 1000000` flag (for the google-closure-compiler package), otherwise the install would get stuck on the last package
2. I would need to install build-essentials (for gcc) - don't know if this was for ubuntu or the entire linux family in general
3. I would require firefox and google chrome to be installed in order for the tests to work

DibyodyutiMondal commented:
Apart from the above, I am at my wit's end

following is the output just for the forms package on my local PC
```
dibyodyuti@FALCON:~/github/angular$ yarn bazel test --cache_test_results=no  //packages/forms/...
yarn run v1.22.17
$ /home/dibyodyuti/github/angular/node_modules/.bin/bazel test --cache_test_results=no //packages/forms/...
DEBUG: /home/dibyodyuti/.cache/bazel/_bazel_dibyodyuti/d4a32c8202ac3fa7a590bc2bd0fa04c7/external/build_bazel_rules_nodejs/index.bzl:73:14: yarn_install#yarn attribute not set and no repository named 'yarn' exists; installing default yarn
INFO: Analyzed 16 targets (2 packages loaded, 890 targets configured).
INFO: Found 10 targets and 6 test targets...
INFO: Elapsed time: 8.783s, Critical Path: 7.51s
INFO: 7 processes: 11 linux-sandbox.
INFO: Build completed successfully, 7 total actions
//packages/forms:forms_api                                               PASSED in 2.6s
//packages/forms:forms_errors                                            PASSED in 1.9s
//packages/forms/test:circular_deps_test                                 PASSED in 0.5s
//packages/forms/test:test                                               PASSED in 3.7s
//packages/forms/test:test_web_chromium                                  PASSED in 5.5s
//packages/forms/test:test_web_firefox                                   PASSED in 6.4s

Executed 6 out of 6 tests: 6 tests pass.
INFO: Build completed successfully, 7 total actions
Done in 8.90s.
dibyodyuti@FALCON:~/github/angular$
```

But if I try to run all the tests, it fails... specifically `/packages/core/test:test_web_chromium` it fails to build and the entire test flow stops
```
ERROR: /home/dibyodyuti/github/angular/packages/core/test/BUILD.bazel:18:8: Executing genrule //packages/core/test:downleveled_es5_fixture failed: (Exit 127): bash failed: error executing command
  (cd /home/dibyodyuti/.cache/bazel/_bazel_dibyodyuti/d4a32c8202ac3fa7a590bc2bd0fa04c7/sandbox/linux-sandbox/1424/execroot/angular && \
  exec env - \
    PATH=/bin:/usr/bin:/usr/local/bin \
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh;
        es2015_out_dir="/tmp/downleveled_es5_fixture/"
        es2015_out_file="$es2015_out_dir/es2015_inheritance_fixture.js"

        # Build the ES2015 output and then downlevel it to ES5.
        bazel-out/host/bin/external/npm/typescript/bin/tsc.sh packages/core/test/reflection/es2015_inheritance_fixture.ts --outDir $es2015_out_dir --target ES2015             --types --module umd
        bazel-out/host/bin/external/npm/typescript/bin/tsc.sh --outFile bazel-out/k8-fastbuild/bin/packages/core/test/reflection/es5_downleveled_inheritance_fixture.js $es2015_out_file --allowJs             --types --target ES5 --downlevelIteration
    ')
# Configuration: af4acc8b121ad5f302c96faa2d22e1c2a2d6a8ff8da0a0327f4f4acc7c1b91df
# Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
/bin/bash: $'\r': command not found
INFO: Elapsed time: 88.819s, Critical Path: 26.54s
INFO: 1996 processes: 1365 internal, 483 linux-sandbox, 12 local, 136 worker.
FAILED: Build did NOT complete successfully
//packages/animations:animations_api                                  NO STATUS
//packages/animations:animations_errors                               NO STATUS
//packages/animations/browser/test:test_web_chromium                  NO STATUS
//packages/animations/browser/test:test_web_firefox                   NO STATUS
//packages/animations/browser/test:testing_circular_deps_test         NO STATUS
//packages/animations/test:circular_deps_test                         NO STATUS
//packages/animations/test:test                                       NO STATUS
//packages/animations/test:test_web_chromium                          NO STATUS
//packages/animations/test:test_web_firefox                           NO STATUS
//packages/bazel/test/ng_package:common_package                       NO STATUS
//packages/bazel/test/ng_package:core_package                         NO STATUS
//packages/bazel/test/ng_package:example_package                      NO STATUS
//packages/bazel/test/ngc-wrapped:flat_module_test                    NO STATUS
//packages/bazel/test/ngc-wrapped:ngc_test                            NO STATUS
//packages/bazel/test/ngc-wrapped/ivy_enabled:ng_module_ivy_test      NO STATUS
//packages/bazel/test/ngc-wrapped/ivy_enabled:test_module_warnings    NO STATUS
//packages/bazel/test/types_bundle:test                               NO STATUS
//packages/benchpress/test:test                                       NO STATUS
//packages/common:common_api                                          NO STATUS
//packages/common:common_errors                                       NO STATUS
//packages/common/http/test:circular_deps_test                        NO STATUS
//packages/common/http/test:test                                      NO STATUS
//packages/common/http/test:test_web_chromium                         NO STATUS
//packages/common/http/test:test_web_firefox                          NO STATUS
//packages/common/http/testing/test:circular_deps_test                NO STATUS
//packages/common/http/testing/test:test                              NO STATUS
//packages/common/http/testing/test:test_web_chromium                 NO STATUS
//packages/common/http/testing/test:test_web_firefox                  NO STATUS
//packages/common/locales:closure_locale_file                         NO STATUS
//packages/common/test:circular_deps_test                             NO STATUS
//packages/common/test:test                                           NO STATUS
//packages/common/test:test_web_chromium                              NO STATUS
//packages/common/test:test_web_firefox                               NO STATUS
//packages/common/test:testing_circular_deps_test                     NO STATUS
//packages/common/upgrade/test:circular_deps_test                     NO STATUS
//packages/common/upgrade/test:test                                   NO STATUS
//packages/compiler:tsec_test                                         NO STATUS
//packages/compiler-cli:compiler_options_api                          NO STATUS
//packages/compiler-cli:error_code_api                                NO STATUS
//packages/compiler-cli:extended_template_diagnostic_name_api         NO STATUS
//packages/compiler-cli/integrationtest/bazel/injectable_def/app/test:test NO STATUS
//packages/compiler-cli/integrationtest/bazel/injector_def/ivy_build/app/test:test NO STATUS
//packages/compiler-cli/linker/babel/test:test                        NO STATUS
//packages/compiler-cli/linker/test:test                              NO STATUS
//packages/compiler-cli/ngcc/test:integration                         NO STATUS
//packages/compiler-cli/ngcc/test:test                                NO STATUS
//packages/compiler-cli/src/ngtsc/annotations/common/test:test        NO STATUS
//packages/compiler-cli/src/ngtsc/annotations/component/test:test     NO STATUS
//packages/compiler-cli/src/ngtsc/annotations/directive/test:test     NO STATUS
//packages/compiler-cli/src/ngtsc/annotations/ng_module/test:test     NO STATUS
//packages/compiler-cli/src/ngtsc/annotations/test:test               NO STATUS
//packages/compiler-cli/src/ngtsc/core/test:test                      NO STATUS
//packages/compiler-cli/src/ngtsc/cycles/test:test                    NO STATUS
//packages/compiler-cli/src/ngtsc/entry_point/test:test               NO STATUS
//packages/compiler-cli/src/ngtsc/file_system/test:test               NO STATUS
//packages/compiler-cli/src/ngtsc/file_system/testing/test:test       NO STATUS
//packages/compiler-cli/src/ngtsc/imports/test:test                   NO STATUS
//packages/compiler-cli/src/ngtsc/incremental/test:test               NO STATUS
//packages/compiler-cli/src/ngtsc/indexer/test:test                   NO STATUS
//packages/compiler-cli/src/ngtsc/logging/test:test                   NO STATUS
//packages/compiler-cli/src/ngtsc/metadata/test:test                  NO STATUS
//packages/compiler-cli/src/ngtsc/partial_evaluator/test:test         NO STATUS
//packages/compiler-cli/src/ngtsc/reflection/test:test                NO STATUS
//packages/compiler-cli/src/ngtsc/scope/test:test                     NO STATUS
//packages/compiler-cli/src/ngtsc/shims/test:test                     NO STATUS
//packages/compiler-cli/src/ngtsc/sourcemaps/test:test                NO STATUS
//packages/compiler-cli/src/ngtsc/transform/test:test                 NO STATUS
//packages/compiler-cli/src/ngtsc/translator/test:test                NO STATUS
//packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/invalid_banana_in_box:test NO STATUS
//packages/compiler-cli/src/ngtsc/typecheck/test:test                 NO STATUS
//packages/compiler-cli/src/ngtsc/util/test:test                      NO STATUS
//packages/compiler-cli/test:downlevel_decorator_transform            NO STATUS
//packages/compiler-cli/test:extract_i18n                             NO STATUS
//packages/compiler-cli/test:perform_compile                          NO STATUS
//packages/compiler-cli/test:perform_watch                            NO STATUS
//packages/compiler-cli/test:typescript_support                       NO STATUS
//packages/compiler-cli/test:version_helpers                          NO STATUS
//packages/compiler-cli/test/compliance/full:full                     NO STATUS
//packages/compiler-cli/test/compliance/linked:linked                 NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/class_metadata.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/components_and_directives.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/components_and_directives/content_projection.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/components_and_directives/lifecycle_hooks.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/components_and_directives/pipes.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/components_and_directives/queries.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/components_and_directives/standalone.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/components_and_directives/template_variables.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/components_and_directives/value_composition.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/elements.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_compiler_compliance/ng_modules.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler/animations.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler/any.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler/hello_world.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler/interpolations.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler/nullish_coalescing.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler/safe_access.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_bindings/attribute_bindings.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_bindings/host_bindings.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_bindings/non_bindable_behavior.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_bindings/property_bindings.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_bindings/text_bindings.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_di/di.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_directives/matching.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/element_attributes.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/es5_support.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/icu_logic.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/line_ending_normalization.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/localize_legacy_message_ids.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/namespaces.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/nested_nodes.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/ng-container_ng-template.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/self-closing_i18n_instructions.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_i18n/whitespace_preserving_mode.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_input_outputs.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_listener.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_providers.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/chaining.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/component_animations.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/component_styles.golden NO STATUS
//packages/compiler-cli/test/compliance/test_cases:source_mapping/external_templates.golden NO STATUS
//packages/compiler-cli/test/ngtsc:ngtsc                              NO STATUS
//packages/compiler/test:circular_deps_test                           NO STATUS
//packages/compiler/test/expression_parser:expression_parser          NO STATUS
//packages/compiler/test/expression_parser:expression_parser_web_chromium NO STATUS
//packages/compiler/test/ml_parser:ml_parser                          NO STATUS
//packages/compiler/test/ml_parser:ml_parser_web_chromium             NO STATUS
//packages/compiler/test/render3:test                                 NO STATUS
//packages/compiler/test/selector:selector                            NO STATUS
//packages/compiler/test/selector:selector_web_chromium               NO STATUS
//packages/compiler/test/selector:selector_web_firefox                NO STATUS
//packages/core:base_locale_file                                      NO STATUS
//packages/core:core_api                                              NO STATUS
//packages/core:core_errors                                           NO STATUS
//packages/core:ng_global_utils_api                                   NO STATUS
//packages/core:tsec_test                                             NO STATUS
//packages/core/schematics/test:test                                  NO STATUS
//packages/core/schematics/test/google3:google3                       NO STATUS
//packages/core/src/compiler:tsec_test                                NO STATUS
//packages/core/src/di/interface:tsec_test                            NO STATUS
//packages/core/src/interface:tsec_test                               NO STATUS
//packages/core/src/util:tsec_test                                    NO STATUS
//packages/core/test:circular_deps_test                               NO STATUS
//packages/core/test:test                                             NO STATUS
//packages/core/test:test_web_firefox                                 NO STATUS
//packages/core/test:testing_circular_deps_test                       NO STATUS
//packages/core/test/acceptance:acceptance                            NO STATUS
//packages/core/test/acceptance:acceptance_web_chromium               NO STATUS
//packages/core/test/acceptance:acceptance_web_firefox                NO STATUS
//packages/core/test/bundling/animations:symbol_test                  NO STATUS
//packages/core/test/bundling/animations:test                         NO STATUS
//packages/core/test/bundling/cyclic_import:symbol_test               NO STATUS
//packages/core/test/bundling/cyclic_import:test                      NO STATUS
//packages/core/test/bundling/forms_reactive:symbol_test              NO STATUS
//packages/core/test/bundling/forms_reactive:test                     NO STATUS
//packages/core/test/bundling/forms_template_driven:symbol_test       NO STATUS
//packages/core/test/bundling/forms_template_driven:test              NO STATUS
//packages/core/test/bundling/hello_world:symbol_test                 NO STATUS
//packages/core/test/bundling/hello_world:test                        NO STATUS
//packages/core/test/bundling/hello_world_r2:test                     NO STATUS
//packages/core/test/bundling/injection:symbol_test                   NO STATUS
//packages/core/test/bundling/injection:test                          NO STATUS
//packages/core/test/bundling/router:symbol_test                      NO STATUS
//packages/core/test/bundling/router:test                             NO STATUS
//packages/core/test/bundling/standalone_bootstrap:symbol_test        NO STATUS
//packages/core/test/bundling/standalone_bootstrap:test               NO STATUS
//packages/core/test/bundling/todo:symbol_test                        NO STATUS
//packages/core/test/bundling/todo:test                               NO STATUS
//packages/core/test/bundling/todo_i18n:test                          NO STATUS
//packages/core/test/bundling/todo_r2:test                            NO STATUS
//packages/core/test/compiler:compiler                                NO STATUS
//packages/core/test/compiler:compiler_web_chromium                   NO STATUS
//packages/core/test/compiler:compiler_web_firefox                    NO STATUS
//packages/core/test/render3:render3                                  NO STATUS
//packages/core/test/render3:render3_web_chromium                     NO STATUS
//packages/core/test/render3:render3_web_firefox                      NO STATUS
//packages/core/test/render3/ivy:ivy                                  NO STATUS
//packages/core/test/render3/perf:perf                                NO STATUS
//packages/core/test/strict_types:strict_types                        NO STATUS
//packages/elements:elements_api                                      NO STATUS
//packages/elements/test:circular_deps_test                           NO STATUS
//packages/elements/test:test_firefox                                 NO STATUS
//packages/examples/common:protractor_tests_chromium                  NO STATUS
//packages/examples/core:protractor_tests_chromium                    NO STATUS
//packages/examples/forms:protractor_tests_chromium                   NO STATUS
//packages/examples/service-worker/push:protractor_tests_chromium     NO STATUS
//packages/examples/service-worker/registration-options:protractor_tests_chromium NO STATUS
//packages/examples/upgrade/static/ts/full:full_protractor_chromium   NO STATUS
//packages/examples/upgrade/static/ts/lite:lite_protractor_chromium   NO STATUS
//packages/examples/upgrade/static/ts/lite-multi:lite-multi_protractor_chromium NO STATUS
//packages/examples/upgrade/static/ts/lite-multi-shared:lite-multi-shared_protractor_chromium NO STATUS
//packages/forms/test:circular_deps_test                              NO STATUS
//packages/forms/test:test                                            NO STATUS
//packages/forms/test:test_web_chromium                               NO STATUS
//packages/forms/test:test_web_firefox                                NO STATUS
//packages/language-service/test:test                                 NO STATUS
//packages/language-service/test/legacy:legacy                        NO STATUS
//packages/localize:localize_api                                      NO STATUS
//packages/localize/src/localize/test:test                            NO STATUS
//packages/localize/src/utils/test:test                               NO STATUS
//packages/localize/test:circular_deps_test                           NO STATUS
//packages/localize/test:test                                         NO STATUS
//packages/localize/tools/test/migrate/integration:integration        NO STATUS
//packages/localize/tools/test/translate/integration:integration      NO STATUS
//packages/misc/angular-in-memory-web-api/test:test_web_chromium      NO STATUS
//packages/misc/angular-in-memory-web-api/test:test_web_firefox       NO STATUS
//packages/platform-browser:platform-browser_api                      NO STATUS
//packages/platform-browser:tsec_test                                 NO STATUS
//packages/platform-browser-dynamic:platform-browser-dynamic_api      NO STATUS
//packages/platform-browser-dynamic/test:circular_deps_test           NO STATUS
//packages/platform-browser-dynamic/test:test_web_chromium            NO STATUS
//packages/platform-browser-dynamic/test:test_web_firefox             NO STATUS
//packages/platform-browser/animations:tsec_test                      NO STATUS
//packages/platform-browser/animations/test:test_web_chromium         NO STATUS
//packages/platform-browser/animations/test:test_web_firefox          NO STATUS
//packages/platform-browser/test:circular_deps_test                   NO STATUS
//packages/platform-browser/test:test                                 NO STATUS
//packages/platform-browser/test:test_web_chromium                    NO STATUS
//packages/platform-browser/test:test_web_firefox                     NO STATUS
//packages/platform-browser/test:testing_circular_deps_test           NO STATUS
//packages/platform-server:platform-server_api                        NO STATUS
//packages/platform-server:tsec_test                                  NO STATUS
//packages/platform-server/init/test:circular_deps_test               NO STATUS
//packages/platform-server/init/test:test                             NO STATUS
//packages/platform-server/test:circular_deps_test                    NO STATUS
//packages/platform-server/test:test                                  NO STATUS
//packages/router:router_api                                          NO STATUS
//packages/router/test:circular_deps_test                             NO STATUS
//packages/router/test:test                                           NO STATUS
//packages/router/test:test_web_chromium                              NO STATUS
//packages/router/test:test_web_firefox                               NO STATUS
//packages/router/test:testing_circular_deps_test                     NO STATUS
//packages/router/upgrade/test:circular_deps_test                     NO STATUS
//packages/router/upgrade/test:test_web_chromium                      NO STATUS
//packages/router/upgrade/test:test_web_firefox                       NO STATUS
//packages/service-worker:service-worker_api                          NO STATUS
//packages/service-worker/test:circular_deps_test                     NO STATUS
//packages/service-worker/test:test                                   NO STATUS
//packages/service-worker/test:test_web_chromium                      NO STATUS
//packages/service-worker/test:test_web_firefox                       NO STATUS
//packages/service-worker/worker/test:circular_deps_test              NO STATUS
//packages/upgrade:upgrade_api                                        NO STATUS
//packages/upgrade/src/common/test:circular_deps_test                 NO STATUS
//packages/upgrade/src/common/test:test_chromium                      NO STATUS
//packages/upgrade/src/common/test:test_firefox                       NO STATUS
//packages/upgrade/src/dynamic/test:test_chromium                     NO STATUS
//packages/upgrade/src/dynamic/test:test_firefox                      NO STATUS
//packages/upgrade/static/test:test_chromium                          NO STATUS
//packages/upgrade/static/test:test_firefox                           NO STATUS
//packages/upgrade/static/testing/test:test_chromium                  NO STATUS
//packages/zone.js/test:browser_disable_wrap_uncaught_promise_rejection_karma_jasmine_test_chromium NO STATUS
//packages/zone.js/test:browser_green_test_karma_jasmine_test_chromium NO STATUS
//packages/zone.js/test:browser_legacy_test_karma_jasmine_test_chromium NO STATUS
//packages/zone.js/test:browser_shadydom_karma_jasmine_test_chromium  NO STATUS
//packages/zone.js/test:browser_test_karma_jasmine_test_chromium      NO STATUS
//packages/zone.js/test:karma_jasmine_test_ci_chromium                NO STATUS
//packages/zone.js/test:test_node_error_disable_policy                NO STATUS
//packages/zone.js/test:test_node_error_lazy_policy                   NO STATUS
//packages/zone.js/test:test_node_no_jasmine_clock                    NO STATUS
//packages/zone.js/test:test_npm_package                              NO STATUS
//packages/core/test:test_web_chromium                          FAILED TO BUILD
//packages/animations/browser/test:circular_deps_test                    PASSED in 0.9s
//packages/animations/browser/test:test                                  PASSED in 3.0s
//packages/common:base_currencies_file                                   PASSED in 0.2s
//packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/nullish_coalescing_not_nullable:test PASSED in 3.1s
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/binding_slots.golden PASSED in 0.5s
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/class_bindings.golden PASSED in 0.3s
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/host_bindings.golden PASSED in 0.3s
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/interpolations.golden PASSED in 0.3s
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/invalid.golden PASSED in 0.3s
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/mixed_style_and_class.golden PASSED in 0.4s
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_styling/style_bindings.golden PASSED in 0.3s
//packages/compiler-cli/test/compliance/test_cases:r3_view_compiler_template.golden PASSED in 0.4s
//packages/compiler-cli/test/compliance/test_cases:source_mapping/inline_templates.golden PASSED in 0.3s
//packages/compiler/test:test                                            PASSED in 4.6s
//packages/compiler/test:test_web_chromium                               PASSED in 9.2s
//packages/compiler/test:test_web_firefox                                PASSED in 11.3s
//packages/compiler/test/expression_parser:expression_parser_web_firefox PASSED in 11.3s
//packages/compiler/test/ml_parser:ml_parser_web_firefox                 PASSED in 11.0s
//packages/core/src/reflection:tsec_test                                 PASSED in 4.9s
//packages/elements/test:test_chromium                                   PASSED in 5.9s
//packages/examples/core:test                                            PASSED in 2.1s
//packages/examples/core/testing/ts:test                                 PASSED in 1.7s
//packages/forms:forms_api                                               PASSED in 6.3s
//packages/forms:forms_errors                                            PASSED in 5.6s
//packages/localize/schematics/ng-add:test                               PASSED in 2.8s
//packages/localize/tools/test:test                                      PASSED in 5.6s
//packages/localize/tools/test/extract/integration:integration           PASSED in 5.0s
//packages/platform-browser-dynamic/test:test                            PASSED in 2.1s
//packages/platform-browser/animations/test:circular_deps_test           PASSED in 1.0s
//packages/platform-browser/animations/test:test                         PASSED in 2.9s
//packages/platform-server/init:tsec_test                                PASSED in 7.0s
//packages/platform-server/test:testing_circular_deps_test               PASSED in 2.3s
//packages/service-worker/config/test:circular_deps_test                 PASSED in 1.2s
//packages/service-worker/config/test:test                               PASSED in 5.5s
//packages/service-worker/worker/test:test                               PASSED in 1.3s
//packages/upgrade/static/test:circular_deps_test                        PASSED in 0.9s
//packages/upgrade/static/testing/test:circular_deps_test                PASSED in 1.5s
//packages/upgrade/static/testing/test:test_firefox                      PASSED in 11.3s
//packages/zone.js/test:test_node                                        PASSED in 7.9s
//packages/zone.js/test:test_node_bluebird                               PASSED in 1.5s

Executed 40 out of 302 tests: 40 tests pass, 1 fails to build and 261 were skipped.
FAILED: Build did NOT complete successfully
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

DibyodyutiMondal commented:
@dylhunn 
It seems that the build and lint issues shown by circle ci on github have been resolved, but some checks are still failing.
I am not aware of how pullapprove works. Is there something I have to on my end?

dylhunn commented:
Let me rerun it -- it looks like CI lint and test didn't run

DibyodyutiMondal committed:
feat(forms): add an input property to form directives that makes ngSubmit wait for async validation to complete before it emits

fix #31021

Before this change, ngSubmit would emit immediately after running sync validation, even though async validation was still running. Thus, any code that handled ngSubmit had no guarantees that the form data was in fact valid.

After this change, adding submitAfterAsyncValidation to the form tag in the template will indicate the form directive to wait for async validation to complete before it emits.

DibyodyutiMondal committed:
fixup! feat(forms): add an input property to form directives that makes ngSubmit wait for async validation to complete before it emits

DibyodyutiMondal commented:
I think I finally managed to do it. 

The original issue with my pull request was that the goldens for the test were never updated. This was because a few tests were not even building, hence the tests did not run. Thus, it did not generate the log files that had the command for updating the goldens. Once I got the test to build (I used node@16.15.0 of node instead of node@16.14.2), it allowed me to fix it.

And as I was re-writing it all, I introduced a circular dependency.

And now the ci won't run the tests again.

DibyodyutiMondal committed:
fixup! feat(forms): add an input property to form directives that makes ngSubmit wait for async validation to complete before it emits

DibyodyutiMondal commented:
The CI test fails at `//integrations/forms`

So I ran `yarn test //integrations/forms/...` locally
That resulted in the following error (`Permission Denied`) which I have no idea how to resolve
```
/bin/sh: 1: /home/dibyodyuti/.cache/bazel/_bazel_dibyodyuti/d4a32c8202ac3fa7a590bc2bd0fa04c7/execroot/angular/bazel-out/k8-fastbuild/bin/integration/forms/test.sh.runfiles/angular/scripts/ci/bazel-payload-size.sh: Permission denied
Error: Integration test command: `/home/dibyodyuti/.cache/bazel/_bazel_dibyodyuti/d4a32c8202ac3fa7a590bc2bd0fa04c7/execroot/angular/bazel-out/k8-fastbuild/bin/integration/forms/test.sh.runfiles/angular/scripts/ci/bazel-payload-size.sh forms 'dist/*.js' true /home/dibyodyuti/.cache/bazel/_bazel_dibyodyuti/d4a32c8202ac3fa7a590bc2bd0fa04c7/execroot/angular/bazel-out/k8-fastbuild/bin/integration/forms/test.sh.runfiles/angular/goldens/size-tracking/integration-payloads.json` failed. See error output above for details.
    at TestRunner.<anonymous> (external/npm/@angular/dev-infra-private/bazel/integration/test_runner/runner.ts:233:15)
    at Generator.next (<anonymous>)
    at fulfilled (/home/dibyodyuti/.cache/bazel/_bazel_dibyodyuti/d4a32c8202ac3fa7a590bc2bd0fa04c7/execroot/angular/bazel-out/k8-fastbuild/bin/integration/forms/test.sh.runfiles/npm/@angular/dev-infra-private/bazel/integration/test_runner/runner.js:11:58)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```

dylhunn commented:
I'm sorry that we weren't able to help get your configuration issues resolved, and the tests green. If you are able to get it working in the future, we can continue efforts to review and discuss, but since this has been inactive for a while, I'm closing it for now.

DibyodyutiMondal commented:
The changes to the code and tests are really not that complex. Is it not possible for someone who is familiar with committing to the angular repo to just cherry pick the code changes into their own branch, and then run the necessary commands to make the goldens and conduct tests etc? You guys must know some contributor who can help me out here.
